envoy agent connection keepalive

courtlane opened the pull request:
**What this PR does / why we need it**:

This PR introduces the `connection_keepalive` configuration for `http2_protocol_options` in the Envoy-Agent cluster configuration.

**Motivation**

In our setup, we are using a Cilium-managed LoadBalancer IP for `nodeport-proxy` service. This IP is announced via ARP on the control plane node. When Cilium performs a failover and reassigns this IP to another node, it announces the new MAC address via ARP. However, the Envoy agent in the usercluster is unaware that the existing TCP connection is now stale or half-open, due to the ARP-based reassignment.

This leads to scenarios where Envoy continues to use a broken TCP connection, resulting in failed or hanging requests. In practice, this causes the kube-apiserver to become unreachable from the user cluster **until the Envoy agent DaemonSets are manually restarted** to force reconnection.

While the Linux kernel will eventually detect and close stale TCP connections (typically after ~20 minutes, depending on system settings), this delay is not acceptable for production-grade control plane access.



**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #

Envoy provides an HTTP/2 `connection_keepalive` feature that sends periodic HTTP/2 PING frames to verify connection liveness. If no PING ACK is received within the defined `timeout`, Envoy will close the connection and attempt to re-establish it, thereby recovering from stale connections automatically.

https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto.html#envoy-v3-api-msg-config-core-v3-keepalivesettings

```yaml
clusters:
  - name: proxy_cluster
    connect_timeout: 5s
    type: LOGICAL_DNS
    typed_extension_protocol_options:
      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
        "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
        explicit_http_config:
          http2_protocol_options:
            connection_keepalive:
              interval: 2s
              timeout: 1s
```


**What type of PR is this?**

/kind bug

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add HTTP/2 keepalive functionality enabling Envoy-Agent to detect broken or
half-open TCP connections to the nodeport-proxy and re-establish them
automatically
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

courtlane committed:
Add connection keepalive to http2_protocol_options

Signed-off-by: Oleg Franko <o.franko@outlook.de>

courtlane committed:
Update http2 connection_keepalive interval and timout

Signed-off-by: Oleg Franko <o.franko@outlook.de>

cnvergence commented on pkg/controller/user-cluster-controller-manager/resources/resources/envoy-agent/configmap.go:
```diff
@@ -90,7 +90,10 @@ static_resources:
-            http2_protocol_options: {}
+            http2_protocol_options:
+              connection_keepalive:
+                interval: 2s
```
Thanks @Courtlane for the PR, just wanted to ask for the rationale behind these values :)
Wonder if we might have more relaxed settings, as it will be the new default for all environments out there

courtlane replied:
Makes sense, @cnvergence

Since Envoy doesn’t provide clear defaults, the current values are based on practical testing.

We’re open to adjusting. For example, with a 10 second interval and a 5 second timeout, stale connections would be detected and recovered in approximately 15 seconds, which could still meet the goal without much overhead.

Alternatively, would it be possible to make these settings customizable or override them to better suit our specific use case?

cnvergence replied:
I would say adjusting it to have less frequent probes is fine for me :)
We can also make it configurable by adding templates

courtlane replied:
@cnvergence How should we proceed with this?
Would you prefer that we adjust the check interval and timeout values and merge them as hardcoded defaults, or should we go for a cleaner solution by making these settings configurable?

I don’t do much programming, especially in Go, so implementing the templated solution would probably take me quite some time to figure out.

cnvergence replied:
just coming back to this, sorry, I wasn't available
@Courtlane, I see your point. We can do the templating as a follow-up :)
We can merge this PR, but let's adjust the settings

courtlane committed:
Align connection_keepalive values with Kubelb configuration

With these values stale connections will be detected and recovered in
approximately 35 seconds

Signed-off-by: Oleg Franko <o.franko@outlook.de>

ahmedwaleedmalik reviewed (approved):
/approve

/cc @cnvergence

cnvergence reviewed (approved):
Thanks for addressing this
